### ISSUE TITLE
Inconsistent output from `usgs` vs `nwm`

### ISSUE THREAD

**Issue by jarq6c** [RTIInternational/teehr#294]
Comparing what `fetching.usgs` returns:
```console
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   value_time          17514 non-null  datetime64[ms]
 1   location_id         17514 non-null  string        
 2   reference_time      0 non-null      float64       
 3   value               17514 non-null  float32       
 4   variable_name       17514 non-null  category      
 5   unit_name           17514 non-null  string        
 6   configuration_name  17514 non-null  category      
dtypes: category(2), datetime64[ms](1), float32(1), float64(1), string(2)

           value_time    location_id  reference_time     value           variable_name unit_name configuration_name
2 2020-10-01 00:00:00  usgs-02146409             NaN  0.224553  streamflow_hourly_inst     m^3/s  usgs_observations
3 2020-10-01 01:00:00  usgs-02146409             NaN  0.217473  streamflow_hourly_inst     m^3/s  usgs_observations
4 2020-10-01 02:00:00  usgs-02146409             NaN  0.210677  streamflow_hourly_inst     m^3/s  usgs_observations
5 2020-10-01 03:00:00  usgs-02146409             NaN  0.210677  streamflow_hourly_inst     m^3/s  usgs_observations
6 2020-10-01 04:00:00  usgs-02146409             NaN  0.210677  streamflow_hourly_inst     m^3/s  usgs_observations
```
to what `fetching.nwm` returns:
```console
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   value_time          17520 non-null  datetime64[ms]
 1   location_id         17520 non-null  string        
 2   value               17520 non-null  float32       
 3   unit_name           17520 non-null  category      
 4   variable_name       17520 non-null  category      
 5   configuration_name  17520 non-null  category      
 6   reference_time      0 non-null      float64       
dtypes: category(3), datetime64[ms](1), float32(1), float64(1), string(1)

           value_time    location_id  value unit_name variable_name   configuration_name  reference_time
0 2020-10-01 00:00:00  nwm30-9731488   0.20    m3 s-1    streamflow  nwm30_retrospective             NaN
1 2020-10-01 00:00:00  nwm30-9731486   0.94    m3 s-1    streamflow  nwm30_retrospective             NaN
2 2020-10-01 01:00:00  nwm30-9731488   0.20    m3 s-1    streamflow  nwm30_retrospective             NaN
3 2020-10-01 01:00:00  nwm30-9731486   0.94    m3 s-1    streamflow  nwm30_retrospective             NaN
4 2020-10-01 02:00:00  nwm30-9731488   0.20    m3 s-1    streamflow  nwm30_retrospective             NaN2
```
It appears the TEEHR implicit data model isn't applied equitably. Sometimes `streamflow` is `streamflow_hourly_inst`. Sometimes `unit_name` is a category, sometimes it's a string. Both `m^3/s` and `m3 s-1` are used. @christophertubbs may have also seen some instances where `value_time` may return as either `datetime64[ns]` or `datetime64[ms]`.

Is there any consideration for enforcing/validating the TEEHR data model?

**Comment by mgdenno** [comment]
@jarq6c First, we are very happy to have feedback and absolutely willing to adjust our approach.  And probably some of what you are pointing out should be addressed, while other things maybe or maybe not.  

There are multiple ways to fetch data from external sources using TEEHR.  One is, for example if fetching USGS gage data, to call `usgs_to_parquet()` in the `src/teehr/fetching/usgs/usgs.py` module. I think this is what you are doing. The other is to create a new Evaluation object (`ev`) and call `ev.fetch.usgs()`.  These two methods behave differently.  The intent was to keep the former as close to "raw" as possible while the latter passes through `conversion` and is `validated and inserted` into the TEEHR dataset.  The thought was, if users are just using the fetching functions outside of an evaluation they may prefer to have the more raw data while if they are using it within a TEEHR Evaluation, it must conform to our data model, etc.  The intent/hope was that most users would be using TEEHR for the TEEHR Evaluation and would use the latter approach which should not have this issue. Roughly speaking, the `usgs_to_parquet()` function does the "Extract" of the "Extract-Transform-Load" (ETL) process while `ev.fetch.usgs()' does the whole ETL.  That said, it was a little tricky to know where to draw the lines and I can see we have not been completely consistent.  At a minimum it seems ensuring that the `dtypes` are consistent after the "Extract" stage would make sense, and if we were going to keep it "raw" from a values perspective then the USGS `variable_name` should be `00060`.  And if we really want it "raw", then should the `location_id` values have the source prefix, should it even be a Pandas DataFrame... aaaahhh!

While we did consider all of this, we didn't agonize too much over where to draw the lines and are happy to adjust course if we missed the target. I think some of this also ties back to the conversation earlier this week about the future of OWPHydroTools.  If the fetching functions from TEEHR were to be spun off as a separate package, how should it behave... How much "Transform" should a fetching library perform?

Happy to continue the conversation here or jump on a call to discuss if that makes sense.

cc: @samlamont

**Comment by jarq6c** [comment]
Yeah that all makes sense and I have no strong opinions about how the final form should look. It's also possible some of this may be resolved as the documentation is updated, as it sounds like I may have misunderstood how the various TEEHR functions are supposed to work. I think from a user's perspective, a certain level of consistency can make the package as a whole more intuitive to use, but maybe I missed something. Based on your description, I might expect a set of simple uncoupled interfaces that might look like this for usgs data:

1. `usgs_json_to_pandas` (extract)
2. `usgs_pandas_to_teehr` (transform)
3. `load_teehr_data` (load)

Then I might expect some convenience methods with a higher level of coupling, perhaps:

1. `usgs_to_teehr` (extract and transform)
2. `load_usgs` (extract and transform, if needed, or just load)

**Note: I'm not actually proposing these interfaces.** I'm just trying to understand how we might communicate TEEHR convention and expected usage patterns. For example, the `pandas` [Group by: split-apply-combine](https://pandas.pydata.org/docs/user_guide/groupby.html#group-by-split-apply-combine) pattern is fundamental to the use of `pandas` and an important component of any training geared towards teaching people how to get the most out of `pandas`. Another important `pandas` consideration is developing pipelines in such a way that you remove for-loops and make use of NumPy vectorization as much as possible.

What are comparable important considerations for the use of TEEHR?

**Comment by jarq6c** [comment]
There may also be some merit in leaning into the theme of TEEHR by explicating what the various "tiers" are and the methods that correspond to those tiers. So maybe in the docs, there's a "Low Level Methods" page, a "Mid Level Methods" page, and  "High Level Methods" page. Or perhaps the TEEHR interface has other logical divisions comparable to the object ("advanced/low level") and state-based ("easy/high level") interfaces of `matplotlib`.

**Comment by christophertubbs** [comment]
I haven't used the usgs stuff - just retrospective to parquet from `teehr.fetching.nwm.retrospective_points` since that seemed to be the most obvious. I didn't even know there were other approaches.

My problem occurred when I was trying to join the data I was playing with - I'd get a ton of data, perform some transformations on the values (not the metadata columns - those don't change), save the result, rinse and repeat, and when I go to zip everything together, it fails because it can't reconcile datetime[ns] with datetime[us] with datetime[ms]. Same code, same approach, different datatypes.
